### ISSUE TITLE
different behavoir of "typeset" for "declare and assign" and "assign after declare"

### ISSUE THREAD

**Issue by sneyx1234** [att/ast#1312]
**Description of problem:**
Different behavoir with ugly sideeffects if a associative array with compound/subname is declared and implicit a value is assigned and if the same is done in two steps -- first declare and second assign a value.

**Ksh version:**
```
# print -v .sh.version
Version AJM 93u+ 2012-08-01
```

**How reproducible:**
Script o interactve mode.

**Steps to reproduce:**
1.
```
typeset KEY='k1'

  unset    A_ASSO
typeset -A A_ASSO
typeset -p A_ASSO

typeset -A A_ASSO[${KEY}].COMPOUND_SUBNAME="declare_and_assign_noindex_fail"
typeset -p A_ASSO

  unset    B_ASSO
typeset -A B_ASSO
typeset -p B_ASSO

typeset -A B_ASSO[${KEY}].COMPOUND_SUBNAME[0]="declare_and_assign_index_succ"
typeset -p B_ASSO

  unset    C_ASSO
typeset -A C_ASSO
typeset -p C_ASSO

typeset -A C_ASSO[${KEY}].COMPOUND_SUBNAME
           C_ASSO[${KEY}].COMPOUND_SUBNAME="assign_after_declare_noindex_succ"
typeset -p C_ASSO

  unset    D_ASSO
typeset -A D_ASSO
typeset -p D_ASSO

typeset -A D_ASSO[${KEY}].COMPOUND_SUBNAME
           D_ASSO[${KEY}].COMPOUND_SUBNAME[0]="assign_after_declare_index_succ"
typeset -p D_ASSO
```

**Actual results:**
typeset -A A_ASSO=()
typeset -A A_ASSO=(['${KEY}']=(typeset -A COMPOUND_SUBNAME;) [k1]=(COMPOUND_SUBNAME=declare_and_assign_noindex_fail))
typeset -A B_ASSO=()
typeset -A B_ASSO=([k1]=(typeset -a COMPOUND_SUBNAME=(declare_and_assign_index_succ);))
typeset -A C_ASSO=()
typeset -A C_ASSO=([k1]=(typeset -A COMPOUND_SUBNAME=([0]=assign_after_declare_noindex_succ);))
typeset -A D_ASSO=()
typeset -A D_ASSO=([k1]=(typeset -A COMPOUND_SUBNAME=([0]=assign_after_declare_index_succ);))

**Expected results:**
typeset -A A_ASSO=()
typeset -A A_ASSO=([k1]=(COMPOUND_SUBNAME=declare_and_assign_noindex_fail))
typeset -A B_ASSO=()
typeset -A B_ASSO=([k1]=(typeset -a COMPOUND_SUBNAME=(declare_and_assign_index_succ);))
typeset -A C_ASSO=()
typeset -A C_ASSO=([k1]=(typeset -A COMPOUND_SUBNAME=([0]=assign_after_declare_noindex_succ);))
typeset -A D_ASSO=()
typeset -A D_ASSO=([k1]=(typeset -A COMPOUND_SUBNAME=([0]=assign_after_declare_index_succ);))

**Additional info:**

The wrong behavoir is in the A_ASSO case ... a phatom key "'${KEY}'" is created as sideffect.  
The same result if the second "typeset -A" of each case is replaced by "typeset".

Replacing the string type with integer type may help to find the root cause:
Script 2 contents:
```
typeset KEY='k1'

  unset    X_ASSO
typeset -A X_ASSO
typeset -p X_ASSO

integer    X_ASSO[${KEY}].COMPOUND_SUBNAME="111"
typeset -p X_ASSO

  unset    Y_ASSO
typeset -A Y_ASSO
typeset -p Y_ASSO

integer    Y_ASSO[${KEY}].COMPOUND_SUBNAME
           Y_ASSO[${KEY}].COMPOUND_SUBNAME="333"
typeset -p Y_ASSO

```
Result:
```
typeset -A X_ASSO=()
typeset -A X_ASSO=(['${KEY}']=(typeset -l -i COMPOUND_SUBNAME=0) [k1]=(COMPOUND_SUBNAME=111))
typeset -A Y_ASSO=()
typeset -A Y_ASSO=([k1]=(typeset -l -i COMPOUND_SUBNAME=333))
```
Expected:
```
typeset -A X_ASSO=()
typeset -A X_ASSO=([k1]=(typeset -l -i COMPOUND_SUBNAME=111))
typeset -A Y_ASSO=()
typeset -A Y_ASSO=([k1]=(typeset -l -i COMPOUND_SUBNAME=333))
```

If the line in focus is enclosed in a "eval \\"...\\"" the bogus behavoir did not occur ... which may lead to believe the order of variable expansion is for "typeset" somehow reordered  ??

Script 3 contents:
```
typeset KEY='k1'

  unset    X_ASSO
typeset -A X_ASSO
typeset -p X_ASSO

eval "integer    X_ASSO[${KEY}].COMPOUND_SUBNAME=\"111\""
typeset -p X_ASSO

  unset    Y_ASSO
typeset -A Y_ASSO
typeset -p Y_ASSO

integer    Y_ASSO[${KEY}].COMPOUND_SUBNAME
           Y_ASSO[${KEY}].COMPOUND_SUBNAME="333"
typeset -p Y_ASSO


```

**Comment by krader1961** [comment]
I get the expected, correct, result (modulo the missing `typeset -a` for the A_ASSO case in your example) using ksh built from the git master branch. I do see the incorrect result with the ksh93 that ships with macOS. I'll add this as a regression test.

**Comment by sneyx1234** [comment]
I can confirm it is fixed in the current master on solaris 11.4 too. Would be nice to know what was the fix for it. May be it was already in 93v-.

**Comment by krader1961** [comment]
> May be it was already in 93v-.

I confirmed it was fixed in that source version. But since we don't have any commit history worth talking about prior to moving the project to Github we'll never know what was changed to fix this issue.
